src/useClientRequest.js

```diff
@@ -88,7 +88,14 @@ function useClientRequest(query, initialOpts = {}) {
     }
 
     dispatch({ type: actionTypes.LOADING });
-    const result = await client.request(revisedOperation, revisedOpts);
+    let result = await client.request(revisedOperation, revisedOpts);
+
+    if (state.data && result.data && revisedOpts.updateResult) {
+      if (typeof revisedOpts.updateResult !== 'function') {
+        throw new Error('options.updateResult must be a function');
+      }
+      result.data = revisedOpts.updateResult(state.data, result.data);
```

Comment:
Since the name is `updateResult`, maybe we should pass the entire result? It might be useful for a developer to change / override the errors in some circumstances.

`useClientRequest` would then look something like:
```js
let result = await client.request(revisedOperation, revisedOpts);

if (revisedOpts.updateResult) {
  if (typeof revisedOpts.updateResult !== 'function') {
    throw new Error('options.updateResult must be a function');
  }
  result = revisedOpts.updateResult(state, result)
}
```

The pagination example `updateResult`:
```js
const updateResult = (prevResult, newResult) => ({
  ...newResult,
  data: {
    ...newResult.data,
    allPosts: [...prevResult.data.allPosts, ...result.data.allPosts]
  }
});

useQuery(MY_QUERY, {
  updateResult
})
```

It makes updating the data a bit more verbose, but the developer could write a wrapper if it was a really common usecase:

```js
const dataUpdater = (fn) => (prevResult, newResult) => ({
  ...newResult,
  data: fn(prevResult.data, newResult.data)
})
```

Reply:
As discussed, passing the entire result back allows internal properties like `loading` & `error` to be changed, or new ones to be added which could be abused or cause tricky bugs. We're going to rename this to `updateData` as a better description of what it's doing & if we see a use case for changing / overriding errors in the future, we could always introduce an `updateErrors` option.